packages/hardhat-ignition-core/src/track-transaction.ts

```diff
@@ -0,0 +1,251 @@
+
+  /**
+   * Cases to consider:
+   * 1. (happy case) given txhash matches a nonce we prepared but didn't record sending
+   * 2. (user replaced with different tx) given txhash matches a nonce we prepared but didn't record sending,
+   *     but the tx details are different
+   * 3. (user sent known txhash) given txhash matches a nonce we recorded sending with the same txhash
+   * 4. (user sent unknown txhash) given txhash matches a nonce we recorded sending but with a different txhash
+   * 5. (user sent unrelated txhash) given txhash doesn't match any nonce we've allocated
+   */
+  for (const exState of exStates) {
+    for (const networkInteraction of exState.networkInteractions) {
+      if (
+        networkInteraction.type ===
+          NetworkInteractionType.ONCHAIN_INTERACTION &&
+        exState.from.toLowerCase() === transaction.from.toLowerCase() &&
+        networkInteraction.nonce === transaction.nonce
+      ) {
+        if (networkInteraction.transactions.length === 0) {
+          // case 1: the txHash matches a transaction we appear to have sent
+          if (
+            networkInteraction.to?.toLowerCase() ===
+              transaction.to?.toLowerCase() &&
+            networkInteraction.data === transaction.data &&
+            networkInteraction.value === transaction.value
+          ) {
+            let fees: NetworkFees;
+            if (
+              "maxFeePerGas" in transaction &&
+              "maxPriorityFeePerGas" in transaction &&
+              transaction.maxFeePerGas !== undefined &&
+              transaction.maxPriorityFeePerGas !== undefined
+            ) {
+              fees = {
+                maxFeePerGas: transaction.maxFeePerGas,
+                maxPriorityFeePerGas: transaction.maxPriorityFeePerGas,
+              };
+            } else {
+              assertIgnitionInvariant(
+                "gasPrice" in transaction && transaction.gasPrice !== undefined,
+                "Transaction fees are missing"
+              );
+
+              fees = {
+                gasPrice: transaction.gasPrice,
+              };
+            }
+
+            const transactionSendMessage: TransactionSendMessage = {
+              futureId: exState.id,
+              networkInteractionId: networkInteraction.id,
+              nonce: networkInteraction.nonce,
+              type: JournalMessageType.TRANSACTION_SEND,
+              transaction: {
+                hash: transaction.hash,
+                fees,
+              },
+            };
+
+            await applyNewMessageFn(
+              transactionSendMessage,
+              deploymentState,
+              deploymentLoader
+            );
+
+            return;
+          }
+          // case 2: the user sent a different transaction that replaced ours
+          // so we check their transaction for the required number of confirmations
+          else {
+            return checkConfirmations(
```

Comment:
I'm not sure if I understand this part, let's look at it on a call. Other than that LGTM.

Reply:
self-answer: This works because it's the else to the `if` in line 104. Hence, the transaction matches the sender and nonce, but not data and value. i.e. a replacement.